src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/UpsertSearchParamsWithOptimisticConcurrency.sql

```diff
@@ -0,0 +1,81 @@
+﻿/*************************************************************
+    Stored procedures - UpsertSearchParams_2
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     UpsertSearchParams_2
+--
+-- DESCRIPTION
+--     Given a set of search parameters, creates or updates the parameters.
+--     Implements optimistic concurrency control using LastUpdated.
+--
+-- PARAMETERS
+--     @searchParams
+--         * The updated existing search parameters or the new search parameters
+--
+-- RETURN VALUE
+--     The IDs and URIs of the search parameters that were inserted (not updated).
+--
+CREATE PROCEDURE dbo.UpsertSearchParamsWithOptimisticConcurrency
+    @searchParams dbo.SearchParamList READONLY
+AS
+SET NOCOUNT ON;
+SET XACT_ABORT ON;
+SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;
+BEGIN TRANSACTION;
+
+DECLARE @lastUpdated AS DATETIMEOFFSET (7) = SYSDATETIMEOFFSET();
+DECLARE @summaryOfChanges TABLE (
+    Uri    VARCHAR (128) COLLATE Latin1_General_100_CS_AS NOT NULL,
+    Action VARCHAR (20)  NOT NULL);
+
+-- Declare table to collect concurrency conflicts
+DECLARE @conflictedRows TABLE (
+    Uri VARCHAR(128) COLLATE Latin1_General_100_CS_AS NOT NULL
+);
+
+-- Adding WITH TABLOCKX higher up in transaction per PR 5097 review comments
+-- Acquire and hold an exclusive table lock for the entire transaction to prevent parameters from being added or modified during upsert.
+SELECT TOP 0 * FROM dbo.SearchParam WITH (TABLOCKX);
```

Comment:
No need to put tablockx here. You can put it on next statement (insert into conflicts)